Incorrect pointer generation

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Pro v19044 64-bit
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz  

CC version: cc (Rev3, Built by MSYS2 project) 12.1.0

getwd: C:\Users\[redacted]\Documents\v\acta
vmodules: C:\Users\[redacted]\.vmodules
vroot: C:\v
vexe: C:\v\v.exe
vexe mtime: 2022-07-18 08:27:51
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 28fd176.12d57e8

Git version: git version 2.37.1
Git vroot status: weekly.2022.28-29-g12d57e8e
.git/config present: true
thirdparty/tcc: N/A

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg src/assembler2.v`
See attached file `src/assembler2.v`

**What did you expect to see?**

Successful compilation or a useful error message


**What did you see instead?**
```
==================
C:\Users\[redacted]\AppData\Local\Temp\v_0\assembler2.7074951881432286981.tmp.c: In function 'main__Parser_parse':
C:\Users\[redacted]\AppData\Local\Temp\v_0\assembler2.7074951881432286981.tmp.c:18000:100: error: incompatible type for argument 1 of 'array_get_with_check'
18000 |                                         int* _t5 = (int*)/*ee elem_ptr_typ */(array_get_with_check(value, 1));
      |                                                                                                    ^~~~~
      |                                                                                                    |
      |                                                                                                    Array_int * {aka array *}
C:\Users\[redacted]\AppData\Local\Temp\v_0\assembler2.7074951881432286981.tmp.c:7158:52: note: expected 'array' but argument is of type 'Array_int *' {aka 'array *'}
 7158 | VV_LOCAL_SYMBOL voidptr array_get_with_check(array a, int i) {
      |                                              ~~~~~~^
C:\Users\[redacted]\AppData\Local\Temp\v_0\assembler2.7074951881432286981.tmp.c:18039:164: error: expected ')' before '_t9'
18039 |                                         main__Maybe_T_f64 as_float = ((as_int)._typ == 18 /* bool */ ? (                                                _option_f64 _t9 = strconv__atof64(operand);
      |                                                                                                        ~                                                           ^~~~
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
[assembler2.zip](https://github.com/vlang/v/files/9131025/assembler2.zip)
Reading the code, the first error seems to be related to the variable `value`, which is an element of `Parser.statements` which is `[][]int`. Perhaps nested arrays aren't quite handled properly?
No idea what's going on with the second error, though it seems related to my homebrew Maybe type. (If there's now a standard way to do Maybe types that's been added since I last messed with this, please let me know! Option types would seem like the way to do it, but I don't remember them working as needed.)

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -252,6 +252,9 @@ fn (mut g Gen) index_of_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 		tmp_opt_ptr := if gen_or { g.new_tmp_var() } else { '' }
 		if gen_or {
 			g.write('$elem_type_str* $tmp_opt_ptr = ($elem_type_str*)/*ee elem_ptr_typ */(array_get_with_check(')
+			if left_is_ptr && !node.left_type.has_flag(.shared_f) {
+				g.write('*')
+			}
 		} else {
 			if needs_clone {
 				g.write('/*2*/string_clone(')

--- a/vlib/v/tests/array_index_optional.v
+++ b/vlib/v/tests/array_index_optional.v
@@ -4,10 +4,13 @@ mut:
 }
 
 fn test_mut_array_index_optional() {
-	mut arr := Arrs{}
+	mut arr := Arrs{
+		x: [[1, 2]]
+	}
 	for mut sub_arr in arr.x {
-		x := sub_arr[0] or { 666 }
-		assert x == 666
+		x := sub_arr[0] or { 3 }
+		println(x)
+		assert x == 1
 	}
 }
 

--- a/vlib/v/tests/array_index_optional.v
+++ b/vlib/v/tests/array_index_optional.v
@@ -1,0 +1,20 @@
+struct Arrs {
+mut:
+	x [][]int
+}
+
+fn test_mut_array_index_optional() {
+	mut arr := Arrs{}
+	for mut sub_arr in arr.x {
+		x := sub_arr[0] or { 666 }
+		assert x == 666
+	}
+}
+
+fn test_array_index_optional_with_if_expr() {
+	ret := []string{}[0] or {
+		if true { 'a' } else { 'b' }
+	}
+	println(ret)
+	assert ret == 'a'
+}

--- a/vlib/v/tests/array_index_optional_with_if_expr_test.v
+++ b/vlib/v/tests/array_index_optional_with_if_expr_test.v
@@ -1,7 +1,0 @@
-fn test_array_index_optional_with_if_expr() {
-	ret := []string{}[0] or {
-		if true { 'a' } else { 'b' }
-	}
-	println(ret)
-	assert ret == 'a'
-}

